### ISSUE TITLE
code-coverage is badly broken on master

### ISSUE THREAD

**Issue by vtjnash** [JuliaLang/julia#52391]
seems like loading pkgimages with code-coverage flags immediately aborts (because it doesn't have `--pkgimage=no` also)

```
julia> versioninfo()
Julia Version 1.11.0-DEV.1002
Commit bac3ba5697* (2023-11-30 03:12 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 128 × AMD EPYC 7513 32-Core Processor
  WORD_SIZE: 64
  LLVM: libLLVM-15.0.7 (ORCJIT, znver3)
  Threads: 1 on 128 virtual cores
Environment:
  JULIA_PKG_PRECOMPILE_AUTO = 0

$ ./julia --code-coverage=x.info --code-coverage=user -e 'using REPL'
julia: /home/vtjnash/julia/src/staticdata.c:2236: jl_root_new_gvars: Assertion `codeinst && (jl_atomic_load_relaxed(&codeinst->specsigflags) & 0b01) && jl_atomic_load_relaxed(&codeinst->specptr.fptr)' failed.

[588330] signal 6 (-6): Aborted
in expression starting at none:1
pthread_kill at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
raise at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7f44b27e071a)
__assert_fail at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
jl_root_new_gvars at /home/vtjnash/julia/src/staticdata.c:2236 [inlined]
jl_restore_system_image_from_stream_ at /home/vtjnash/julia/src/staticdata.c:3446
jl_restore_package_image_from_stream at /home/vtjnash/julia/src/staticdata.c:3580
jl_restore_incremental_from_buf at /home/vtjnash/julia/src/staticdata.c:3627
ijl_restore_package_image_from_file at /home/vtjnash/julia/src/staticdata.c:3715
_include_from_serialized at ./loading.jl:1073
_include_from_serialized at ./loading.jl:1048 [inlined]
_require_search_from_serialized at ./loading.jl:1626
_require at ./loading.jl:2039
__require_prelocked at ./loading.jl:1916
jfptr___require_prelocked_63295 at /home/vtjnash/julia/usr/lib/julia/sys.so (unknown line)
_jl_invoke at /home/vtjnash/julia/src/gf.c:2906 [inlined]
ijl_apply_generic at /home/vtjnash/julia/src/gf.c:3088
jl_apply at /home/vtjnash/julia/src/julia.h:2139 [inlined]
jl_f__call_in_world at /home/vtjnash/julia/src/builtins.c:888
#invoke_in_world#3 at ./essentials.jl:989 [inlined]
invoke_in_world at ./essentials.jl:986 [inlined]
_require_prelocked at ./loading.jl:1907
macro expansion at ./loading.jl:1845 [inlined]
macro expansion at ./lock.jl:267 [inlined]
__require at ./loading.jl:1806
jfptr___require_63227 at /home/vtjnash/julia/usr/lib/julia/sys.so (unknown line)
_jl_invoke at /home/vtjnash/julia/src/gf.c:2906 [inlined]
ijl_apply_generic at /home/vtjnash/julia/src/gf.c:3088
jl_apply at /home/vtjnash/julia/src/julia.h:2139 [inlined]
jl_f__call_in_world at /home/vtjnash/julia/src/builtins.c:888
#invoke_in_world#3 at ./essentials.jl:989 [inlined]
invoke_in_world at ./essentials.jl:986 [inlined]
require at ./loading.jl:1799
jfptr_require_63224 at /home/vtjnash/julia/usr/lib/julia/sys.so (unknown line)
_jl_invoke at /home/vtjnash/julia/src/gf.c:2906 [inlined]
ijl_apply_generic at /home/vtjnash/julia/src/gf.c:3088
jl_apply at /home/vtjnash/julia/src/julia.h:2139 [inlined]
call_require at /home/vtjnash/julia/src/toplevel.c:484 [inlined]
eval_import_path at /home/vtjnash/julia/src/toplevel.c:521
jl_toplevel_eval_flex at /home/vtjnash/julia/src/toplevel.c:757
jl_toplevel_eval_flex at /home/vtjnash/julia/src/toplevel.c:884
ijl_toplevel_eval_in at /home/vtjnash/julia/src/toplevel.c:992
eval at ./boot.jl:428 [inlined]
exec_options at ./client.jl:291
_start at ./client.jl:525
jfptr__start_64816 at /home/vtjnash/julia/usr/lib/julia/sys.so (unknown line)
_jl_invoke at /home/vtjnash/julia/src/gf.c:2906 [inlined]
ijl_apply_generic at /home/vtjnash/julia/src/gf.c:3088
jl_apply at /home/vtjnash/julia/src/julia.h:2139 [inlined]
true_main at /home/vtjnash/julia/src/jlapi.c:586
jl_repl_entrypoint at /home/vtjnash/julia/src/jlapi.c:738
main at /home/vtjnash/julia/cli/loader_exe.c:58
unknown function (ip: 0x7f44b27e1d8f)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_start at ./julia (unknown line)
Allocations: 8 (Pool: 8; Big: 0); GC: 0
Aborted (core dumped)
134 vtjnash@deepsea4:~/julia$ 
```

**Comment by IanButterworth** [comment]
Something's not right here because nightly is working in the wild with coverage turned on and your example doesn't error for me
```
ian@Ians-MacBook-Pro julia % ./julia --code-coverage=x.info --code-coverage=user -e 'using REPL'
ian@Ians-MacBook-Pro julia %
```
```
julia> versioninfo()
Julia Version 1.11.0-DEV.1010
Commit f7618602d4 (2023-12-04 14:15 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin23.0.0)
  CPU: 10 × Apple M2 Pro
  WORD_SIZE: 64
  LLVM: libLLVM-15.0.7 (ORCJIT, apple-m1)
  Threads: 1 on 6 virtual cores
```

**Comment by vtjnash** [comment]
It is an assert error, which most people don't see

**Comment by vtjnash** [comment]
Fixed by #53373
